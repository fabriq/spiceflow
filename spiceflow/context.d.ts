import * as context from './dist/context';
export = context;
