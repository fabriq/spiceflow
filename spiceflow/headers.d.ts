import * as headers from './dist/headers';
export = headers;
