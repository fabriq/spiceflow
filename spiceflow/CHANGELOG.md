# spiceflow

## 0.0.3

### Patch Changes

- Fix \_\_dirname

## 0.0.2

### Patch Changes

- Added experimental --openapi

## 0.0.1

### Patch Changes

- 0eef621: Initial release
