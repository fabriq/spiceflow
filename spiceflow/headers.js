module.exports = require('./dist/headers');
