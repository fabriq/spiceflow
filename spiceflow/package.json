{
  "name": "spiceflow",
  "version": "1.2.0",
  "description": "Simple API framework with RPC and type safety",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "repository": "https://github.com/remorses/spiceflow",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./cors": {
      "types": "./dist/cors.d.ts",
      "default": "./dist/cors.js"
    },
    "./mcp": {
      "types": "./dist/mcp.d.ts",
      "default": "./dist/mcp.js"
    },
    "./client": {
      "types": "./dist/client/index.d.ts",
      "default": "./dist/client/index.js"
    },
    "./openapi": {
      "types": "./dist/openapi.d.ts",
      "default": "./dist/openapi.js"
    },
    "./dist/*": {
      "types": "./dist/*.d.ts",
      "default": "./dist/*.js"
    }
  },
  "scripts": {
    "build": "cp ../README.md ./README.md && rm -rf dist && tsc",
    "prepublishOnly": "pnpm build",
    "gen-openapi": "pnpm tsx scripts/openapi-fern.ts",
    "docs:dev": "pnpm run gen-openapi && fern docs dev",
    "docs:prod": "pnpm run gen-openapi && fern generate --docs --force",
    "fern-sdk": "pnpm run gen-openapi && fern generate --force ",
    "fern-docs": "pnpm run gen-openapi && fern generate --docs --force",
	"play-sdk:build": "pnpm vite build --config ./scripts/play-sdk.vite.ts",
    "test": "pnpm vitest",
    "prepare": "pnpm build",
    "watch": "tsc -w"
  },
  "files": [
    "dist",
    "src"
  ],
  "keywords": [],
  "author": "Tommaso De Rossi, morse <beats.by.morse@gmail.com>",
  "license": "",
  "dependencies": {
    "@medley/router": "^0.2.1",
    "@modelcontextprotocol/sdk": "^1.0.4",
    "@sinclair/typebox": "^0.33.7",
    "ajv": "^8.17.1",
    "ajv-formats": "^3.0.1",
    "eventsource-parser": "^2.0.1",
    "fast-querystring": "^1.1.2",
    "lodash.clonedeep": "^4.5.0",
    "openapi-types": "^12.1.3",
    "zod": "^3.23.8",
    "zod-to-json-schema": "^3.24.1"
  },
  "devDependencies": {
    "@types/node": "22.7.4",
    "eventsource": "^3.0.2",
    "formdata-node": "^6.0.3",
    "js-base64": "^3.7.7",
    "js-yaml": "^4.1.0"
  }
}
