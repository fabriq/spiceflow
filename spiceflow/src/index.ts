export { Spiceflow } from './spiceflow.js'
export type { AnySpiceflow } from './spiceflow.js'
export { InternalServerError, ParseError, ValidationError } from './error.js'
