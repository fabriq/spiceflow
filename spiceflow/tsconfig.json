{
  "extends": "../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "src",
    "module": "NodeNext",
    "target": "ESNext",
    "lib": ["DOM", "ESNext", "dom.iterable"],
    "moduleResolution": "NodeNext",
    "declarationMap": true,
    "sourceMap": false,
    "resolveJsonModule": true,
    
    "outDir": "dist"
  },
  "include": ["src"]
}
