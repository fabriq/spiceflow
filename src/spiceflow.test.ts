import { expect } from "@std/expect";
import { bfs, cloneDeep, Spiceflow } from "./spiceflow.ts";
import { z } from "zod";
import { createSpiceflowClient } from "./client/mod.ts";
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("works", async () => {
  const res = await new Spiceflow()
    .post("/xxx", () => "hi")
    .handle(new Request("http://localhost/xxx", { method: "POST" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});

Deno.test("cloneDeep", async (t) => {
  await t.step("works on promises", async () => {
    const obj = {
      promise: Promise.resolve({ value: "hi" }),
    };
    const res = cloneDeep(obj);
    expect(res).not.toBe(obj);
    expect(Object.keys(res)).toEqual(Object.keys(obj));
    expect(res.promise).toBeInstanceOf(Promise);
    // expect(res.promise).not.toBe(obj.promise);
    expect(await res.promise).toEqual({ value: "hi" });
  });
});

Deno.test("can encode superjson types", async (t) => {
  const app = new Spiceflow().post("/superjson", () => {
    const item = {
      date: new Date("2025-01-20T18:01:57.852Z"),
      map: new Map([["a", 1]]),
      set: new Set([1, 2, 3]),
      bigint: BigInt(123),
    };
    return { items: Array(2).fill(item) };
  });
  const res = await app.handle(
    new Request("http://localhost/superjson", { method: "POST" }),
  );
  expect(res.status).toBe(200);
  const client = createSpiceflowClient(app);
  await assertSnapshot(t, await client.superjson.post().then((x) => x.data));
  await assertSnapshot(t, await res.json());
});
Deno.test("dynamic route", async () => {
  const res = await new Spiceflow()
    .post("/ids/:id", () => "hi")
    .handle(new Request("http://localhost/ids/xxx", { method: "POST" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});
Deno.test("handler returns url encoded data", async () => {
  const params = new URLSearchParams();
  params.append("name", "test");
  params.append("value", "123");

  const res = await new Spiceflow()
    .post("/form", () => params, {
      type: "application/x-www-form-urlencoded",
    })
    .handle(
      new Request("http://localhost/form", {
        method: "POST",
      }),
    );

  expect(res.status).toBe(200);
  expect(res.headers.get("content-type")).toBe(
    "application/x-www-form-urlencoded",
  );
  const text = await res.text();
  const responseParams = new URLSearchParams(text);
  expect(responseParams.get("name")).toBe("test");
  expect(responseParams.get("value")).toBe("123");
});
Deno.test("GET dynamic route", async () => {
  const res = await new Spiceflow()
    .get("/ids/:id", () => "hi")
    .post("/ids/:id", ({ params: { id } }) => id, {
      params: z.object({ id: z.string() }),
    })
    .handle(new Request("http://localhost/ids/xxx", { method: "GET" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});

Deno.test("onError does not fire on 404", async () => {
  let errorFired = false;
  const app = new Spiceflow()
    .get("/test", () => "Hello")
    .onError(() => {
      errorFired = true;
      return new Response("Error", { status: 500 });
    });

  const res = await app.handle(
    new Request("http://localhost/non-existent", { method: "GET" }),
  );

  expect(res.status).toBe(404);
  expect(errorFired).toBe(false);
  expect(await res.text()).toBe("Not Found");
});

Deno.test("onError fires on validation errors", async () => {
  let errorMessage = "";
  const app = new Spiceflow()
    .post(
      "/test",
      async ({ request }) => {
        await request.json();
        return "Success";
      },
      {
        body: z.object({
          name: z.string(),
        }),
      },
    )
    .onError(({ error }) => {
      errorMessage = error.message;
      return new Response("Error", { status: 400 });
    });

  const res = await app.handle(
    new Request("http://localhost/test", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ name: 1 }), // Invalid type for 'name'
    }),
  );

  expect(res.status).toBe(400);
  expect(errorMessage).toEqual("name: Expected string, received number");
  expect(await res.text()).toEqual("Error");
});

/*
Deno.test("HEAD uses GET route, does not add body", async () => {
  const app = new Spiceflow().get("/ids/:id", () => {
    console.trace("GET");
    return {
      message: "hi",
      length: 10,
    };
  });

  const res = await app.handle(
    new Request("http://localhost/ids/xxx", { method: "HEAD" }),
  );
  expect(res.status).toBe(200);
  // expect(res.headers.get('Content-Length')).toBe('10')
  expect(await res.text()).toBe("");

  // Compare with GET to ensure HEAD is using GET route
  const getRes = await app.handle(
    new Request("http://localhost/ids/xxx", { method: "GET" }),
  );
  expect(getRes.status).toBe(200);
  expect(await getRes.json()).toEqual({ message: "hi", length: 10 });
});
*/

Deno.test("GET with query, untyped", async () => {
  const res = await new Spiceflow()
    .get("/query", ({ query }) => {
      return query.id;
    })
    .handle(new Request("http://localhost/query?id=hi", { method: "GET" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});

Deno.test("GET with query, zod, fails validation", async () => {
  const res = await new Spiceflow()
    .get(
      "/query",
      ({ query }) => {
        return query.id;
      },
      {
        query: z.object({
          id: z.number(),
        }),
      },
    )
    .handle(new Request("http://localhost/query?id=hi", { method: "GET" }));
  expect(res.status).toBe(422);
});

Deno.test("GET with query and zod", async () => {
  const res = await new Spiceflow()
    .get(
      "/query",
      ({ query }) => {
        // @ts-expect-error
        void query.sdfsd;
        return query.id;
      },
      {
        query: z.object({
          id: z.string(),
        }),
      },
    )
    .handle(new Request("http://localhost/query?id=hi", { method: "GET" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});

Deno.test("GET dynamic route, params are typed", async () => {
  const res = await new Spiceflow()
    .get("/ids/:id", ({ params }) => {
      const id = params.id;
      // @ts-expect-error
      params.sdfsd;
      return id;
    })
    .handle(new Request("http://localhost/ids/hi", { method: "GET" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});

Deno.test("GET dynamic route, params are typed with schema", async () => {
  const res = await new Spiceflow()
    .get(
      "/ids/:id",
      ({ params }) => {
        const id = params.id;
        // @ts-expect-error
        params.sdfsd;
        return id;
      },
      {
        params: z.object({
          id: z.string(),
        }),
      },
    )
    .handle(new Request("http://localhost/ids/hi", { method: "GET" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});

Deno.test("missing route is not found", async () => {
  const res = await new Spiceflow()
    .get("/ids/:id", () => "hi")
    .handle(new Request("http://localhost/zxxx", { method: "GET" }));
  expect(res.status).toBe(404);
});
Deno.test("state works", async () => {
  const res = await new Spiceflow()
    .state("id", "")
    .use(({ state }) => {
      state.id = "xxx";
    })
    .get("/get", ({ state: state }) => {
      expect(state.id).toBe("xxx");
    })
    .handle(new Request("http://localhost/get"));
  expect(res.status).toBe(200);
});

Deno.test("body is parsed as json", async () => {
  let body;
  const res = await new Spiceflow()
    .state("id", "")
    .post("/post", async (c) => {
      body = await c.request.json();
      // console.log({request})
      return body;
    })
    .handle(
      new Request("http://localhost/post", {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({ name: "John" }),
      }),
    );
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual({ name: "John" });
});

Deno.test("validate body works, request success", async () => {
  const res = await new Spiceflow()
    .post(
      "/post",
      async ({ request }) => {
        // console.log({request})
        const body = await request.json();
        expect(body).toEqual({ name: "John" });
        return "ok";
      },
      {
        body: z.object({
          name: z.string(),
        }),
      },
    )
    .handle(
      new Request("http://localhost/post", {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({ name: "John" }),
      }),
    );
  expect(res.status).toBe(200);
  expect(await res.text()).toEqual('"ok"');
});

Deno.test("validate body works, request fails", async () => {
  const res = await new Spiceflow()
    .post(
      "/post",
      async ({ request }) => {
        // console.log({request})
        const body = await request.json();
        expect(body).toEqual({ name: "John" });
      },
      {
        body: z.object({
          name: z.string(),
          requiredField: z.string(),
        }),
      },
    )
    .handle(
      new Request("http://localhost/post", {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({ name: "John" }),
      }),
    );
  expect(res.status).toBe(422);
  expect(await res.text()).toEqual(
    JSON.stringify({
      "code": "VALIDATION",
      "status": 422,
      "message": "requiredField: Required",
    }),
  );
});

Deno.test("run use", async () => {
  const res = await new Spiceflow()
    .use(({ request }) => {
      expect(request.method).toBe("HEAD");
      return new Response("ok", { status: 401 });
    })
    .use(({ request }) => {
      expect(request.method).toBe("HEAD");
      return "second one";
    })
    .head("/ids/:id", () => "hi")
    .handle(new Request("http://localhost/ids/zxxx", { method: "HEAD" }));
  expect(res.status).toBe(401);
  expect(await res.text()).toBe("ok");
});

Deno.test("run use", async () => {
  const res = await new Spiceflow()
    .use(({ request }) => {
      expect(request.method).toBe("HEAD");
      return new Response("ok", { status: 401 });
    })
    .use(({ request }) => {
      expect(request.method).toBe("HEAD");
      return "second one";
    })
    .head("/ids/:id", () => "hi")
    .handle(new Request("http://localhost/ids/zxxx", { method: "HEAD" }));
  expect(res.status).toBe(401);
  expect(await res.text()).toBe("ok");
});

Deno.test("basPath works", async () => {
  const res = await new Spiceflow({ basePath: "/one" })
    .get("/ids/:id", () => "hi")
    .handle(new Request("http://localhost/one/ids/xxx", { method: "GET" }));
  expect(res.status).toBe(200);
  expect(await res.json()).toEqual("hi");
});

Deno.test("basPath works with use", async () => {
  const app = new Spiceflow({ basePath: "/one" }).use(
    new Spiceflow({})
      .get("/two", () => "hi")
      .use(new Spiceflow({ basePath: "/three" }).get("/four", () => "hi")),
  );
  {
    const res = await app.handle(
      new Request("http://localhost/one/two", { method: "GET" }),
    );

    expect(res.status).toBe(200);
    expect(await res.json()).toEqual("hi");
  }
  {
    const res = await app.handle(
      new Request("http://localhost/one/three/four", { method: "GET" }),
    );
    expect(res.status).toBe(200);
    expect(await res.json()).toEqual("hi");
  }
});

Deno.test("getRouteAndParents", async () => {
  const app = new Spiceflow({ basePath: "/one" })
    .get("/ids/:id", () => "hi")
    .use(
      new Spiceflow({ basePath: "/two" }).use(
        new Spiceflow({ basePath: "/three" }).get("/four", () => "hi"),
      ),
    );

  const routers = bfs(app);
  const last = routers[routers.length - 1];

  expect(app["getAppAndParents"](last).map((x) => x.prefix))
    .toEqual(["/one", "/two", "/three"]);
});

Deno.test("getAppsInScope include all parent apps", async () => {
  const app = new Spiceflow({ basePath: "/one" })
    .get("/ids/:id", () => "hi")
    .use(
      new Spiceflow({ basePath: "/two" }).use(
        new Spiceflow({ basePath: "/three" }).use(
          new Spiceflow({ basePath: "/four" })
            .get("/five", () => "hi")
            .use(() => {}),
        ),
      ),
    );

  const routers = bfs(app);
  const secondLast = routers[routers.length - 2];

  expect(app["getAppsInScope"](secondLast).map((x) => x.prefix))
    .toEqual(["/one", "/two", "/three"]);
});

Deno.test("getAppsInScope include all parent apps and non scoped apps", async () => {
  const app = new Spiceflow({ basePath: "/one" })
    .get("/ids/:id", () => "hi")
    .use(
      new Spiceflow({ basePath: "/two" }).use(
        new Spiceflow({ basePath: "/three" }).use(
          new Spiceflow({ basePath: "/four", scoped: false })
            .get("/five", () => "hi")
            .use(() => {}),
        ),
      ),
    );

  const routers = bfs(app);
  const secondLast = routers[routers.length - 2];

  expect(app["getAppsInScope"](secondLast).map((x) => x.prefix))
    .toEqual(["/one", "/two", "/three", "/four"]);
});

Deno.test("use with 2 basPath works", async () => {
  let oneOnReq = false;
  let twoOnReq = false;
  const onReqCalled: string[] = [];
  const app = new Spiceflow()
    .use(() => {
      onReqCalled.push("root");
    })
    .use(
      new Spiceflow({ basePath: "/one" })
        .use(() => {
          oneOnReq = true;
          onReqCalled.push("one");
        })
        .get("/ids/:id", ({ params }) => params.id),
    )
    .use(
      new Spiceflow({ basePath: "/two" })
        .use(() => {
          twoOnReq = true;
          onReqCalled.push("two");
        })
        .get("/ids/:id", ({ params }) => params.id, {}),
    );

  {
    const res = await app.handle(new Request("http://localhost/one/ids/one"));
    expect(res.status).toBe(200);

    expect(await res.json()).toEqual("one");
  }
  expect(onReqCalled).toEqual(["root", "one"]);
  {
    const res = await app.handle(new Request("http://localhost/two/ids/two"));
    expect(res.status).toBe(200);

    expect(await res.json()).toEqual("two");
  }
  expect(oneOnReq).toBe(true);
  expect(twoOnReq).toBe(true);
});

Deno.test("use with nested basPath works", async () => {
  const app = await new Spiceflow({ basePath: "/zero" })
    .use(
      new Spiceflow({ basePath: "/one" }).get(
        "/ids/:id",
        ({ params }) => params.id,
      ),
    )
    .use(
      new Spiceflow({ basePath: "/two" }).use(
        new Spiceflow({ basePath: "/nested" }).get(
          "/ids/:id",
          ({ params }) => params.id,
        ),
      ),
    );
  {
    const res = await app.handle(
      new Request("http://localhost/zero/one/ids/one"),
    );
    expect(res.status).toBe(200);
    expect(await res.json()).toEqual("one");
  }

  {
    const res = await app.handle(
      new Request("http://localhost/zero/two/nested/ids/nested"),
    );
    expect(res.status).toBe(200);
    expect(await res.json()).toEqual("nested");
  }
});

Deno.test("errors inside basPath works", async () => {
  const onErrorTriggered = [] as string[];
  const onReqTriggered = [] as string[];
  let handlerCalledNTimes = 0;
  const app = new Spiceflow({ basePath: "/zero" })
    .onError(() => {
      onErrorTriggered.push("root");
      // return new Response('root', { status: 500 })
    })
    .use(() => {
      onReqTriggered.push("root");
      // return new Response('root', { status: 500 })
    })
    .use(
      new Spiceflow({ basePath: "/two" })
        .onError(() => {
          onErrorTriggered.push("two");
          // return new Response('two', { status: 500 })
        })
        .use(() => {
          onReqTriggered.push("two");
          // return new Response('two', { status: 500 })
        })
        .use(
          new Spiceflow({ basePath: "/nested" })
            .onError(() => {
              onErrorTriggered.push("nested");
              // return new Response('nested', { status: 500 })
            })
            .use(() => {
              onReqTriggered.push("nested");
              // return new Response('nested', { status: 500 })
            })
            .get("/ids/:id", () => {
              handlerCalledNTimes++;
              throw new Error("error message");
            }),
        ),
    );

  {
    const res = await app.handle(
      new Request("http://localhost/zero/two/nested/ids/nested"),
    );
    expect(handlerCalledNTimes).toBe(1);
    expect(onErrorTriggered).toEqual(["root", "two", "nested"]);
    expect(onReqTriggered).toEqual(["root", "two", "nested"]);
    expect(res.status).toBe(500);
    expect(await res.text()).toEqual('{"message":"error message"}');
    // expect(await res.json()).toEqual('nested'))
  }
});

Deno.test("basepath with root route", async () => {
  let handlerCalled = false;
  const app = new Spiceflow({ basePath: "/api" }).get("/", () => {
    handlerCalled = true;
    return new Response("Root route of API");
  });

  const res = await app.handle(new Request("http://localhost/api"));
  expect(handlerCalled).toBe(true);
  expect(res.status).toBe(200);
  expect(await res.text()).toBe("Root route of API");

  // Test that non-root paths are not matched
  handlerCalled = false;
  const res2 = await app.handle(new Request("http://localhost/api/other"));
  expect(handlerCalled).toBe(false);
  expect(res2.status).toBe(404);
});

Deno.test("Trailing slashes and base paths", async (t) => {
  await t.step(
    "App without trailing slash, request with trailing slash",
    async () => {
      const app = new Spiceflow({ basePath: "/api" }).get(
        "/users",
        () => new Response("Users list"),
      );

      const res = await app.handle(new Request("http://localhost/api/users/"));
      expect(res.status).toBe(200);
      expect(await res.text()).toBe("Users list");
    },
  );

  await t.step(
    "App with trailing slash, request without trailing slash",
    async () => {
      const app = new Spiceflow({ basePath: "/api/" }).get(
        "/users/",
        () => new Response("Users list"),
      );

      const res = await app.handle(new Request("http://localhost/api/users"));
      expect(res.status).toBe(200);
      expect(await res.text()).toBe("Users list");
    },
  );

  await t.step("Nested routes with and without trailing slashes", async () => {
    const app = new Spiceflow({ basePath: "/api" }).use(
      new Spiceflow()
        .get("/products", () => new Response("Products list"))
        .get("/categories/", () => new Response("Categories list")),
    );

    const resProducts = await app.handle(
      new Request("http://localhost/api/products/"),
    );
    expect(resProducts.status).toBe(200);
    expect(await resProducts.text()).toBe("Products list");

    const resCategories = await app.handle(
      new Request("http://localhost/api/categories"),
    );
    expect(resCategories.status).toBe(200);
    expect(await resCategories.text()).toBe("Categories list");
  });
});

Deno.test("async generators handle non-ASCII characters correctly", async () => {
  const app = new Spiceflow()
    .get("/cyrillic", async function* () {
      yield "Привет"; // Hello in Russian
      yield "Κόσμος"; // World in Greek
    })
    .get("/mixed-scripts", async function* () {
      // Mix of Cyrillic and Greek letters that look like Latin
      yield { text: "РΡ" }; // Cyrillic and Greek P
      yield { text: "ΟО" }; // Greek and Cyrillic O
      yield { text: "КΚ" }; // Cyrillic and Greek K
    });

  const client = createSpiceflowClient(app);

  const { data: cyrillicData } = await client.cyrillic.get();
  let cyrillicText = "";
  for await (const chunk of cyrillicData!) {
    cyrillicText += chunk;
  }
  expect(cyrillicText).toBe("ПриветΚόσμος");

  const { data: mixedData } = await client["mixed-scripts"].get();
  const mixedResults = [] as any[];
  for await (const chunk of mixedData!) {
    mixedResults.push(chunk);
  }
  expect(mixedResults).toEqual([{ text: "РΡ" }, { text: "ΟО" }, {
    text: "КΚ",
  }]);
});

Deno.test("can pass additional props to body schema", async () => {
  const app = new Spiceflow().post("/user", ({ request }) => request.json(), {
    body: z
      .object({
        name: z.string(),
        age: z.number(),
        email: z.string().email(),
      })
      .passthrough(),
  });

  const res = await app.handle(
    new Request("http://localhost/user", {
      method: "POST",

      body: JSON.stringify({
        name: "John",
        age: 25,
        email: "john@example.com",
        additionalProp: "extra data",
      }),
    }),
  );

  expect(res.status).toBe(200);
  expect(await res.json()).toEqual({
    name: "John",
    age: 25,
    email: "john@example.com",
    additionalProp: "extra data",
  });
});
