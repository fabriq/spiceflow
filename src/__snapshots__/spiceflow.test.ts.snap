export const snapshot = {};

snapshot[`can encode superjson types 1`] = `
{
  items: [
    {
      bigint: 123n,
      date: 2025-01-20T18:01:57.852Z,
      map: Map(1) {
        "a" => 1,
      },
      set: Set(3) {
        1,
        2,
        3,
      },
    },
    {
      bigint: 123n,
      date: 2025-01-20T18:01:57.852Z,
      map: Map(1) {
        "a" => 1,
      },
      set: Set(3) {
        1,
        2,
        3,
      },
    },
  ],
}
`;

snapshot[`can encode superjson types 2`] = `
{
  __superjsonMeta: {
    referentialEqualities: {
      "items.0": [
        "items.1",
      ],
    },
    values: {
      "items.0.bigint": [
        "bigint",
      ],
      "items.0.date": [
        "Date",
      ],
      "items.0.map": [
        "map",
      ],
      "items.0.set": [
        "set",
      ],
      "items.1.bigint": [
        "bigint",
      ],
      "items.1.date": [
        "Date",
      ],
      "items.1.map": [
        "map",
      ],
      "items.1.set": [
        "set",
      ],
    },
  },
  items: [
    {
      bigint: "123",
      date: "2025-01-20T18:01:57.852Z",
      map: [
        [
          "a",
          1,
        ],
      ],
      set: [
        1,
        2,
        3,
      ],
    },
    {
      bigint: "123",
      date: "2025-01-20T18:01:57.852Z",
      map: [
        [
          "a",
          1,
        ],
      ],
      set: [
        1,
        2,
        3,
      ],
    },
  ],
}
`;
